stdlib/regex.gr

```diff
@@ -0,0 +1,2985 @@
+ * - `Z` - Union of `Zl`, `Zp`, and `Zs`
+ * - `Cc` - Other, control
+ * - `Cf` - Other, format
+ * - `Cs` - Other, surrogate
+ * - `Cn` - Other, not assigned
+ * - `Co` - Other, private use
+ * - `C` - Union of `Cc`, `Cf`, `Cs`, `Cn`, and `Co`
+ * - `.` - Union of all Unicode categories
+ *
+ * @param regexString: The regular expression to compile
+ * @returns The compiled regular expression
+ */
+export let make = (regexString: String) => {
+  let buf = makeRegExBuf(regexString)
+  match(parseRegex(buf)) {
+    Err(e) => Err(e),
+    Ok(parsed) => {
+      let numGroups = unbox(buf.config.groupNumber)
+      let references = unbox(buf.config.references)
+      match(validate(parsed, numGroups)) {
+        Err(e) => Err(e),
+        Ok(maxLookbehind) => {
+          let matcher = compileRegexToMatcher(parsed)
+          Ok({
+            reParsed: parsed,
+            reNumGroups: numGroups,
+            reReferences: references,
+            reMaxLookbehind: maxLookbehind,
+            reCompiled: matcher,
+            reMustString: mustString(parsed),
+            reIsAnchored: isAnchored(parsed),
+            reStartRange: startRange(parsed),
+          })
+        }
+      }
+    }
+  }
+}
+
+
+//
+//
+// ============
+// REGEX SEARCH
+// ============
+//
+//
+
+// speed up failures using must-string
+let checkMustString = (ms, buf: MatchBuf, pos, endPos) => {
+  match(ms) {
+    None => true,
+    Some(ms) => {
+      let toCheck = if (pos == 0 && endPos == Array.length(buf.matchInputExploded)) {
+        buf.matchInput
+      } else {
+        String.slice(pos, endPos, buf.matchInput)
+      }
+      Option.isSome(String.indexOf(ms, toCheck))
+    }
+  }
+}
+
+// speed up failures using start-range
+let checkStartRange = (startRange, buf, pos, endPos) => {
+  rangeContains(startRange, Char.code(buf.matchInputExploded[pos]))
+}
+
+
+let searchMatch = (rx: RegularExpression, buf: MatchBuf, pos, startPos, endPos, state) => {
+  if (!checkMustString(rx.reMustString, buf, pos, endPos)) {
+    None
+  } else {
+    let matcher = rx.reCompiled
+    let anchored = rx.reIsAnchored
+    let startRange = rx.reStartRange
+    let rec loop = (pos) => {
+      if (anchored && pos != startPos) {
+        None
+      } else {
+        match(startRange) {
+          Some(_) when pos == endPos => None, // Can't possibly match if chars are required and we are at EOS
+          Some(rng) when !checkStartRange(rng, buf, pos, endPos) => loop(pos + 1),
+          _ => {
+            let pos2 = interp(matcher, buf, pos, startPos, endPos, state)
+            match (pos2) {
+              Some(p) => Some((pos, p)),
+              None when pos < endPos => loop(pos + 1),
+              None => None
+            }
+          }
+        }
+      }
+    }
+    loop(pos)
+  }
+}
+
+/**
+ * The user-facing object which contains the results
+ * of a regular expression match.
+ */
+export record MatchResult {
+  /**
+   * Returns the contents of the given group
+   */
```

Comment:
Remember that we lose these comments in the doc generator.

Reply:
Yep, manually writing them in now